### PR TITLE
Can patch linksToMany fields via AI panel

### PR DESCRIPTION
**Pull request by burieberry**
- [x] Adding item(s) to empty field
- [x] Adding item(s) to field with preexisting values
- [x] Removing item(s) from field (based on position in the list, or based on card title)
- [x] Removing all items
- [x] Replacing an item
- [x] Combo actions: adding/removing/replacing multiple attributes and relationships

Not in scope:
- Nested relationships (CS-6799) - Added nested fields in `trip-info` card, but these aren't fully handled in this PR
- Changing a linked card's inner data (CS-6781)

TODO next: 
Ended up not using the `isPatchApplied` method for relationships patches. It needs to be updated to handle them. Also need to backfill unit tests.

Issues:
- Sometimes AI bot tries to use an incorrect URL. When clicking on Apply,`Patch failed. data not iterable` message appears. Not sure how to ensure that the bot uses correct urls.
- AI bot can be inconsistent with the patch response...

### CODE REVIEW

**Comment by habdelra** [comment]
For the IsPatchApplied, let’s use the new `PatchData` type that I just merged in. It reveals some bugs (that you are probably aware of) in this method.

File: packages/host/app/services/card-service.ts [hunk 1]
```diff
@@ -269,7 +269,16 @@ export default class CardService extends Service {
...
+    // TODO: update isPatchApplied to handle relationships
+    if (
+      patchData.attributes &&
+      Object.keys(patchData.attributes).length &&
+      !this.isPatchApplied(
+        updatedCardDoc.data.attributes,
+        patchData.attributes,
+        card.id,
+      )
+    ) {
```
Comment (habdelra) [inline packages/host/app/services/card-service.ts]: if `isPatchApplied` does not know about relationships, does that mean the apply patch button does the right thing?

Also, deeper question for team--if we are moving to a command based system here, and "patch" is just one type of ai-bot command (e.g. search will be another), how do we deal with  command specific logic in our card service, or in the host in general? probaly we need a way to organize this--and hopefully domain specific command logic has a nice home to live (like the system prompts for commands that live in skill cards).

File: packages/host/app/services/operator-mode-state-service.ts [hunk 2]
```diff
@@ -112,7 +112,42 @@ export default class OperatorModeStateService extends Service {
...
       if ('card' in item && item.card.id == id) {
         let document = await this.cardService.serializeCard(item.card);
-        document.data = merge(document.data, patch);
+        if (patch.attributes) {
+          for (let key of Object.keys(patch.attributes)) {
+            if (!(key in item.card)) {
+              throw new Error(`Attribute "${key}" does not exist in card`);
+            }
+          }
+          document.data.attributes = merge(
```
Comment (habdelra) [inline packages/host/app/services/operator-mode-state-service.ts]: i don't know if this necessarily does the right thing--for instance if the ai bot is trying to remove a property in the patch, does `merge()` actually remove it?

Comment (habdelra) [reply packages/host/app/services/operator-mode-state-service.ts]: take a look at how we handle patches on the server in `Realm.patchCard()`, we use `mergeWith()` which i think is probably what we want here. (the test case is actually patching arrays)

File: packages/host/app/services/operator-mode-state-service.ts [hunk 3]
```diff
@@ -112,7 +112,42 @@ export default class OperatorModeStateService extends Service {
...
+              Object.entries(document.data.relationships)
+                .filter(([fieldName]) => fieldName.includes('.'))
+                .map(([fieldName]) => {
+                  if (fieldName.split('.')[0] === key) {
+                    delete document.data.relationships![fieldName];
+                  }
+                });
+            }
+          }
+          document.data.relationships = merge(
```
Comment (habdelra) [inline packages/host/app/services/operator-mode-state-service.ts]: on the server we are also using `mergeWith` for relationship patching as well, in fact there is a runtime-common function for merging relationships that we use from teh server, `@cardstack/runtime-common/merge-relationships`, is there a reason why we wouldn't use that here as well?

Comment (burieberry) [reply packages/host/app/services/operator-mode-state-service.ts]: Made the updates to use `mergeWith` and `mergeRelationships`. Now this works much better.

**Review by habdelra — APPROVED** [review]

**Pull request merged**